src/librustc/infer/error_reporting/mod.rs

```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
         }
     }
 }
+
+/// Visitor that collects every `Item` imported into the current `TyCtxt`'s scope.
+struct ImportVisitor<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    /// All the `Item`s from other scopes visible from this `TyCtxt`, as well as their local name.
+    full_imports: FxHashMap<String, String>,
+}
+
+impl<'a, 'gcx, 'tcx> ImportVisitor<'a, 'gcx, 'tcx> {
```

Comment:
This looks likely to be buggy - you are reimplementing half of name resolution. Could name resolution be made to supply a `types_in_scope` map?

Reply:
cc @jseyfried - is there a way for resolution to be called after it is complete?

Reply:
I figured that was the case, but it was the closest I've been in any of my approaches to get this working. Once I saw that the suggestion machinery used a visitor it gave me enough context to get this. Having this also allowed me to identify the unit tests that would need to change. It is definitely less than ideal and fully expect there to be a way to do this using existing datastructures, but I feel it is much easier for me to point at this code and saw "this is what I want to do, is there a better way?". :)

Reply:
There's not a way to do name resolution after lowering to HIR (i.e. after we drop the `Resolver`). A naive solution would be to keep the `Resolver` for longer, but that would hog memory and isn't ideal since the Resolver has needless complexity used for import resolution and is based on the AST.

@nikomatsakis and I have discussed creating and saving some name resolution structures for this purpose before destroying the `Resolver`, but we haven't made detailed designs.